### DISCUSSION TITLE
How to install brew tools from custom taps?

### DISCUSSION THREAD

**Discussion by notcool11** [jdx/mise#995] [General]
Hey folks, trying to install `otel-desktop-viewer`, the doc says to do this:
```
brew tap CtrlSpice/homebrew-otel-desktop-viewer
brew install otel-desktop-viewer
```

I can do this in one line like this, which seems to work:
```
brew install ctrlspice/otel-desktop-viewer/otel-desktop-viewer
```

But if I try to put that in my `.rtx.toml` file I get an error:
```
rtx error reading config: /Users/xxx/gitrepos/xxx/.rtx.toml                                                                                                                     
TOML parse error at line 4, column 10
  |
4 | ctrlspice/otel-desktop-viewer/otel-desktop-viewer = "latest"
  |          ^
expected `.`, `=`
```

Any help appreciated, thanks!

**Comment by Marukome0743** [comment]
The immediate TOML parsing error is caused by the `/` characters in an unquoted TOML key. Keys containing `/` must be quoted:

```toml
"ctrlspice/otel-desktop-viewer/otel-desktop-viewer" = "latest"
```

Quoting the key only fixes the TOML syntax, however; it does not automatically make the entry a Homebrew package.

Current mise has declarative Homebrew package support using fully qualified names:

```toml
[bootstrap.packages]
"brew:ctrlspice/otel-desktop-viewer/otel-desktop-viewer" = "latest"
```

I checked the current `CtrlSpice/homebrew-otel-desktop-viewer` tap. It is still active, but it does not currently publish the `api/formula/otel-desktop-viewer.json` metadata required by mise’s native third-party tap support. Therefore, this particular formula cannot currently be installed through `[bootstrap.packages]`.

For this tap, the practical solution remains to use Homebrew directly:

```sh
brew install ctrlspice/otel-desktop-viewer/otel-desktop-viewer
```

This can also be kept in a `Brewfile` or invoked from a mise bootstrap task if it needs to be reproducible.

The current third-party tap requirements are documented here:

https://mise.en.dev/bootstrap/packages/brew.html

*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*

**Reply by jacobbednarz** [reply]
i'm not sure if i'm missing something however, this is working for some third party taps but not others. in both cases, neither are supplying the API metadata; is there some nuance i'm missing here?

working examples:

- https://github.com/tobi/try
- https://github.com/jacobbednarz/homebrew-tap

not working example:

- https://github.com/CtrlSpice/homebrew-otel-desktop-viewer

**Reply by jacobbednarz** [reply]
actually i've found the difference - it's formula vs casks and the latter has a different code path.

**Reply by Marukome0743** [reply]
@jacobbednarz — your diagnosis was right, and my reply above was wrong in a way that matters: I called `otel-desktop-viewer` a formula. It is not. `CtrlSpice/homebrew-otel-desktop-viewer` contains exactly one definition, `Casks/otel-desktop-viewer.rb`, so the manager is `brew-cask:`, not `brew:`. That is the "different code path" you found.

Two things have changed since, and as of **v2026.9.5** both are in a release:

1. **Taps without API metadata are evaluated directly.** mise fetches the Ruby definition at a pinned tap commit and evaluates it with its own Formula/Cask DSL shim (#12774, with `Formula/` → `HomebrewFormula/` → repo-root discovery from #12837). The "must publish `api/…json`" requirement my reply described no longer holds.
2. **On macOS that evaluation was broken until 2026.9.5.** It runs Ruby inside a Seatbelt sandbox, and the profile denied the metadata lookups `realpath` needs on the directories above every allowed path — so Ruby died in `rb_check_realpath_internal` before evaluating anything, for both the Ruby mise provisions and Homebrew's own vendored one. Fixed in #13039 (reported as #12969).

So this now works:

```toml
[bootstrap.packages]
"brew-cask:ctrlspice/otel-desktop-viewer/otel-desktop-viewer" = "latest"
```

Verified today on 2026.9.5 (macOS arm64): `mise bootstrap packages status` fetches `Casks/otel-desktop-viewer.rb` at commit `135ca28c`, evaluates it under `sandbox-exec`, and reports the cask as `missing` — i.e. resolved, not yet installed. On 2026.9.3 the same command ended in `Errno::EPERM`.

What to know:

- Evaluating a tap definition needs **Ruby 3+**. `mise bootstrap packages apply` provisions one if none is found; `status` and `upgrade` do not, and say so.
- Formulae resolved this way are **built from source**; casks download the artifact as usual.
- Only **GitHub-hosted taps** can be fetched directly. For a tap whose GitHub URL cannot be inferred from `owner/tap`, add it under `[bootstrap.brew.taps]` (`mise bootstrap packages brew tap owner/name https://github.com/owner/homebrew-name.git`).

Docs: https://mise.jdx.dev/bootstrap/packages/brew.html#third-party-taps

*AI-assisted — Tool: Claude Code; model: anthropic/claude-opus-5; version: 2.1.236.*

**Reply by jacobbednarz** [reply]
thanks for the update here @Marukome0743 🙇 i checked on the latest version and can confirm this is now working as expected. appreciate the fix on this one!
